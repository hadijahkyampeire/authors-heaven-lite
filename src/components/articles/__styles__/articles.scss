$primary-green: #2ecc71;

.articles-page {
  margin: 1rem 2rem;
  padding: 1rem;

  h2 {
    margin-bottom: 1rem;
    margin-left: 6rem;
  }
  .add-article-button {
    position: fixed;
    bottom: 3rem;
    right: 9rem;
    svg {
      width: 2rem;
      height: 2rem;
      fill: $primary-green;
    }
  }

  .tiles-container {
    display: flex;
    flex-wrap: wrap;
    width: 85%;
    margin: 1rem auto;
    border-top: 1px solid grey;

    .article-tile {
      padding: 1rem 0;
      border-bottom: 1px solid $primary-green;
      width: 100%;

      .des-and-body {
        display: flex;
        flex-wrap: wrap;
        color: inherit;
        text-decoration: none;
      }
      .tile-title {
        margin-bottom: 0.5rem;
        font-weight: bold;
        color: rgba(41, 41, 41, 1);
      }
      .slug {
        font-style: italic;
        color: gray;
        margin-bottom: 1rem;
      }

      .tile-footer {
        display: flex;
        margin-top: 1rem;
        align-items: center;

        .author {
          color: gray;
          text-transform: capitalize;
          margin-right: 1rem;
        }
        .view-more-btn {
          background-color: $primary-green;
          color: white;
          padding: 0.5rem 0.75rem;
          border: none;
          border-radius: 1rem;
          text-decoration: none;
        }
      }
    }
  }
}
