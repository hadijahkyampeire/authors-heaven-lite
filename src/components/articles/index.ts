export * from './addArticles';
export * from './article';
export * from './articles';
export * from './editArticle';
