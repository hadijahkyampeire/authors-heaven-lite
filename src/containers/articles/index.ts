export * from './AddArticle';
export * from './article';
export * from './articles';
export * from './published';
export * from './publishedArticle';
export * from './editArticle';

