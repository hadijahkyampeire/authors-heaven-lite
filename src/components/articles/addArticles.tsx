import React, { useState } from 'react';
import { ApiArticle } from 'types/articles';
import { createArticle } from 'actions/articles';
import { AddArticleForm } from 'components/forms';
import { Redirect } from 'react-router';
import { ArticleData } from 'types/articles';
import { Header } from "components/commons";

import './__styles__/addArticle.scss';

interface Props {
  onSubmit: (
    ...args: Parameters<typeof createArticle>
  ) => void;
  newArticle?: ApiArticle;
}
export const AddArticle: React.FC<Props> = ({ newArticle, onSubmit }) => {
  const initialValues: ArticleData = {
    title: "",
    description: "",
    body: "",
    favourited: false,
    published: false
    
  };

  const [form, setForm] = useState<ArticleData>(initialValues);

  const publishedForm = {...form, published: true };
  const handlePublish = (e: any) => { e.preventDefault(); onSubmit(publishedForm); };
  const handleSave = (e: any) => { e.preventDefault(); onSubmit(form); };
 
  if (newArticle) {
    return <Redirect to={newArticle.published === true ? '/articles/published/' : '/articles/'} />
  }
  return (
    <section className="add-article-page">
      <Header title='Create Article'/>
      <AddArticleForm newArticle={newArticle} onPublish={handlePublish} onSave={handleSave} form={form} setForm={setForm} />
    </section>
  );
};
