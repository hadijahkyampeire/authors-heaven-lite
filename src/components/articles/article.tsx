import { useLocation } from "react-router";
import { Link } from 'react-router-dom';

import { getSingleArticle, deleteArticle } from 'actions/articles';
import React, { useEffect, useState } from 'react';
import { ApiArticle } from 'types/articles';
import { Header, DialogModal } from "components/commons";

import './__styles__/article.scss';

interface Props {
  article?: ApiArticle;
  fetchArticle: (...args: Parameters<typeof getSingleArticle>) => void;
  deleteArticle: (...args: Parameters<typeof deleteArticle>) => void;
  email?: string;
};

export const Article: React.FC<Props> = ({ article, fetchArticle, email, deleteArticle }) => {
  const [openDelete, setOpenDelete] = useState<string | undefined>(undefined);
  const { pathname: url } = useLocation();

  useEffect(() => {
    fetchArticle(url);
  }, [url]); // eslint-disable-line react-hooks/exhaustive-deps

  const hasPermission = email === article?.author?.email;
  return (
    <section className="article-page">
      <Header title={`Viewing ${article?.title }`}/>
      <div>
        <div className="article-title">{article?.title}</div>
        <div className="other-header"><span className="author">By: {article?.author?.username}</span><span>{new Date().getDate()}</span><span>2 min read</span></div>
        <div className="article-body">
          <div className="article-description">{article?.description}</div>
          <div className="article-body">{article?.body}</div>
        </div>
        {hasPermission ? (
          <>
            <div className="article-footer">
              <Link to={`${url}/edit`}>Edit Article</Link>
              <button onClick={() => setOpenDelete(article?.slug)}>Delete Article</button>
              <button>Share Article</button>
            </div>
          </>
          )
        : null}
        <DialogModal 
          open={openDelete === article?.slug} 
          closeModal={() => setOpenDelete(undefined)}
          onSubmit={() => deleteArticle(article?.slug)}
          title={`Delete ${article?.title}`} 
          text='Are you sure you want to delete this Article?' 
           />
      </div>
    </section>
  );
};
