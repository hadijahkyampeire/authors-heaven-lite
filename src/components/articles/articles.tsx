import React, { useEffect } from 'react';
import { useLocation } from "react-router";
import { Link } from 'react-router-dom';

import { DataTile } from 'components/commons';
import { ApiArticlesState } from 'types/articles';
import { AddIcon } from 'components/icons';
import { ContainerTabs } from 'components/commons';
import './__styles__/articles.scss';
import { fetchArticles, deleteArticle } from 'actions/articles';

interface Props {
  articles: ApiArticlesState;
  fetchArticles: (...args: Parameters<typeof fetchArticles>) => void;
  deleteArticle: (...args: Parameters<typeof deleteArticle>) => void;
  email?: string;
}
export const Articles: React.FC<Props> = ({ articles = {}, fetchArticles, deleteArticle, email }) => {
  const { count, results } = articles;
  const { pathname: url } = useLocation();
  useEffect(() => {
    fetchArticles(url)
  }, [fetchArticles, url]);

  const getCount = (tabUrl: string) => url === tabUrl ? count : 0
  const tabslist = [
    {name: `All Articles (${getCount('/articles/')})`, url: '/articles/'},
    {name: `Published Articles  (${getCount('/articles/published/')})`, url: '/articles/published/'}
  ];

  return (
    <section className="articles-page">
      <ContainerTabs tabslist={tabslist} currentPageUrl={url} />
      <div className="tiles-container">
      {results && results.length 
        ? results.map(article => <DataTile article={article} key={article.slug} handleDelete={() => deleteArticle(article.slug)} email={email} />)
        : <div className="no-articles">No Articles Found, Please add some</div>}
      </div>
      <Link to='/articles/new/' className="add-article-button"><AddIcon /></Link>
    </section>
  );
};
