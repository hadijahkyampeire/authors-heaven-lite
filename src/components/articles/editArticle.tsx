import React, { useState, useEffect } from 'react';
import { useLocation } from 'react-router';
import { ApiArticle } from 'types/articles';
import { updateArticle, getSingleArticle } from 'actions/articles';
import { AddArticleForm } from 'components/forms';
import { Redirect } from 'react-router';
import { ArticleData } from 'types/articles';
import { Header } from "components/commons";

import './__styles__/addArticle.scss';

interface Props {
  onUpdate: (...args: Parameters<typeof updateArticle>) => void;
  fetchArticle: (...args: Parameters<typeof getSingleArticle>) => void;
  resetEditArticle: () => void;
  updatedArticle?: ApiArticle;
  article?: ApiArticle
}
export const EditArticle: React.FC<Props> = ({ updatedArticle, onUpdate, article, fetchArticle, resetEditArticle }) => {
  const { pathname: url } = useLocation();
  const initialValues: ArticleData = {
    title: "",
    description: "",
    body: "",
    favourited: false,
    published: false
  };

  const articleUrl = url.replace('/edit', '');

  useEffect(() => resetEditArticle, []); // eslint-disable-line

  useEffect(() => {
    fetchArticle(articleUrl);
  }, [url, fetchArticle]); // eslint-disable-line

  const [form, setForm] = useState<ArticleData>({...initialValues, ...article });

  useEffect(() => setForm({...initialValues, ...article }), [article]); // eslint-disable-line

  const publishedForm = {...form, published: true };

  const handlePublish = (e: any) => { e.preventDefault(); onUpdate(article?.slug, publishedForm ); };
  const handleSave = (e: any) => { e.preventDefault(); onUpdate(article?.slug, form); };

  if (updatedArticle) {
    return <Redirect to='/articles/' />
  }
  return (
    <section className="add-article-page">
      <Header title={`Editing ${article?.title }`}/>
      <AddArticleForm onSave={handleSave} onPublish={handlePublish} form={form} setForm={setForm} context='Edit' />
    </section>
  );
};
