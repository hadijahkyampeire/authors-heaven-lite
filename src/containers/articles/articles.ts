import { connect } from 'react-redux';

import { fetchArticles } from 'actions/articles';
import { Articles } from 'components/articles';
import { AppState } from 'types';


const mapStateToProps = (state: AppState) => ({
  articles: state.articles.articles
});
const mapDispatchToProps = { fetchArticles };

export const ArticlesPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(Articles);
