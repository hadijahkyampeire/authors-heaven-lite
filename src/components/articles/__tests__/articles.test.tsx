import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "testUtils";

import { Articles } from '../articles';

type ArticlesProps = React.ComponentProps<typeof Articles>;

const baseProps: ArticlesProps = {
  fetchArticles: jest.fn(),
  deleteArticle: jest.fn(),
  articles: { 
    results: [{title: '1 one', description: 'desc', body: 'body', slug: '1-one'}, {title: '2 two', description: 'desc', body: 'body', slug: '2-two'}], 
    count: 2 
  }
};

function renderUI(props: Partial<ArticlesProps> = {}) {
  return render(
    <Router>
      <Articles {...baseProps} {...props} />
    </Router>
  );
};

describe('<Articles />', () => {
  it("snapshot test", () => {
    const { container } = renderUI({});
    expect(container).toMatchSnapshot();
  });

  it('renders properly', async () => {
    const { findByTestId, findAllByTestId } = renderUI({});

    const tabContainer = await findByTestId('tabs');
    expect(tabContainer).toBeInTheDocument();

    const dataTile = await findAllByTestId('data-tile');
    expect(dataTile).toHaveLength(2)
  });
  
  it('Shows no article text if there is no article', () => {
    const fewProps: ArticlesProps = {
      fetchArticles: jest.fn(),
      deleteArticle: jest.fn(),
      articles: { 
        results: [], 
        count: 0 
      }
    };
    function renderUI(props: Partial<ArticlesProps> = {}) {
      return render(
        <Router>
          <Articles {...fewProps} {...props} />
        </Router>
      );
    };

    const { getByText } = renderUI({});
    expect(getByText('No Articles Found, Please add some')).toBeInTheDocument();
  });

});
