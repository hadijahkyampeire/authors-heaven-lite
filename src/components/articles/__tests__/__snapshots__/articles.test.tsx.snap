// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles /> snapshot test 1`] = `
<div>
  <section
    class="articles-page"
  >
    <div
      class="tabs"
      data-testid="tabs"
    >
      <a
        class="tab"
        href="/articles/"
        id="All Articles (0)"
      >
        All Articles (0)
      </a>
      <a
        class="tab"
        href="/articles/published/"
        id="Published Articles  (0)"
      >
        Published Articles  (0)
      </a>
    </div>
    <div
      class="tiles-container"
    >
      <div
        class="article-tile"
        data-testid="data-tile"
      >
        <div
          class="tile-title"
        >
          1 one
        </div>
        <div
          class="tile-content"
        >
          <a
            class="des-and-body"
            href="/1-one"
          >
            <div
              class="description"
            >
              desc
              .
            </div>
            <div
              class="body"
            >
              body
            </div>
          </a>
        </div>
        <div
          class="tile-footer"
        >
          <div
            class="publisher"
          >
            <span>
              By: 
            </span>
            <span
              class="author"
            />
          </div>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="open and close list of options"
            class="bx--overflow-menu"
            data-floating-menu-container="true"
            type="button"
          >
            <svg
              aria-label="open and close list of options"
              class="bx--overflow-menu__icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="16"
                cy="8"
                r="2"
              />
              <circle
                cx="16"
                cy="16"
                r="2"
              />
              <circle
                cx="16"
                cy="24"
                r="2"
              />
              <title>
                open and close list of options
              </title>
            </svg>
          </button>
          <div
            class="bx--modal dialog-modal"
            role="presentation"
          >
            <span
              class="bx--visually-hidden"
              role="link"
              tabindex="0"
            >
              Focus sentinel
            </span>
            <div
              aria-label="Delete 1 one"
              aria-modal="true"
              class="bx--modal-container"
              role="dialog"
              tabindex="-1"
            >
              <div
                class="bx--modal-header"
              >
                <button
                  aria-label="Close"
                  class="bx--modal-close"
                  title="Close"
                  type="button"
                >
                  <svg
                    aria-label="Close"
                    class="bx--modal-close__icon"
                    fill="currentColor"
                    focusable="false"
                    height="20"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 32 32"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                    />
                  </svg>
                </button>
                
                <h3
                  class="bx--modal-header__heading"
                  id="bx--modal-header__heading--modal-1"
                >
                  Delete 1 one
                </h3>
              </div>
              <div
                aria-labelledby="bx--modal-header__heading--modal-1"
                class="bx--modal-content"
                id="bx--modal-body--modal-1"
              >
                Are you sure you want to delete this Article?
                <div
                  class="bx--modal-footer bx--btn-set"
                >
                  <button
                    class="bx--btn bx--btn--secondary bx--btn bx--btn--primary"
                    tabindex="0"
                    type="button"
                  >
                    No
                  </button>
                  <button
                    class="bx--btn bx--btn--primary bx--btn bx--btn--primary"
                    tabindex="0"
                    type="button"
                  >
                    Yes
                  </button>
                </div>
              </div>
            </div>
            <span
              class="bx--visually-hidden"
              role="link"
              tabindex="0"
            >
              Focus sentinel
            </span>
          </div>
        </div>
      </div>
      <div
        class="article-tile"
        data-testid="data-tile"
      >
        <div
          class="tile-title"
        >
          2 two
        </div>
        <div
          class="tile-content"
        >
          <a
            class="des-and-body"
            href="/2-two"
          >
            <div
              class="description"
            >
              desc
              .
            </div>
            <div
              class="body"
            >
              body
            </div>
          </a>
        </div>
        <div
          class="tile-footer"
        >
          <div
            class="publisher"
          >
            <span>
              By: 
            </span>
            <span
              class="author"
            />
          </div>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="open and close list of options"
            class="bx--overflow-menu"
            data-floating-menu-container="true"
            type="button"
          >
            <svg
              aria-label="open and close list of options"
              class="bx--overflow-menu__icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="16"
                cy="8"
                r="2"
              />
              <circle
                cx="16"
                cy="16"
                r="2"
              />
              <circle
                cx="16"
                cy="24"
                r="2"
              />
              <title>
                open and close list of options
              </title>
            </svg>
          </button>
          <div
            class="bx--modal dialog-modal"
            role="presentation"
          >
            <span
              class="bx--visually-hidden"
              role="link"
              tabindex="0"
            >
              Focus sentinel
            </span>
            <div
              aria-label="Delete 2 two"
              aria-modal="true"
              class="bx--modal-container"
              role="dialog"
              tabindex="-1"
            >
              <div
                class="bx--modal-header"
              >
                <button
                  aria-label="Close"
                  class="bx--modal-close"
                  title="Close"
                  type="button"
                >
                  <svg
                    aria-label="Close"
                    class="bx--modal-close__icon"
                    fill="currentColor"
                    focusable="false"
                    height="20"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 32 32"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                    />
                  </svg>
                </button>
                
                <h3
                  class="bx--modal-header__heading"
                  id="bx--modal-header__heading--modal-2"
                >
                  Delete 2 two
                </h3>
              </div>
              <div
                aria-labelledby="bx--modal-header__heading--modal-2"
                class="bx--modal-content"
                id="bx--modal-body--modal-2"
              >
                Are you sure you want to delete this Article?
                <div
                  class="bx--modal-footer bx--btn-set"
                >
                  <button
                    class="bx--btn bx--btn--secondary bx--btn bx--btn--primary"
                    tabindex="0"
                    type="button"
                  >
                    No
                  </button>
                  <button
                    class="bx--btn bx--btn--primary bx--btn bx--btn--primary"
                    tabindex="0"
                    type="button"
                  >
                    Yes
                  </button>
                </div>
              </div>
            </div>
            <span
              class="bx--visually-hidden"
              role="link"
              tabindex="0"
            >
              Focus sentinel
            </span>
          </div>
        </div>
      </div>
    </div>
    <a
      class="add-article-button"
      href="/articles/new/"
    >
      <svg
        id="Capa_1"
        version="1.1"
        viewBox="0 0 477.867 477.867"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <g>
            <path
              d="M392.533,0h-307.2C38.228,0.056,0.056,38.228,0,85.333v307.2c0.056,47.105,38.228,85.277,85.333,85.333h307.2 c47.105-0.056,85.277-38.228,85.333-85.333v-307.2C477.81,38.228,439.638,0.056,392.533,0z M443.733,392.533 c0,28.277-22.923,51.2-51.2,51.2h-307.2c-28.277,0-51.2-22.923-51.2-51.2v-307.2c0-28.277,22.923-51.2,51.2-51.2h307.2 c28.277,0,51.2,22.923,51.2,51.2V392.533z"
            />
          </g>
        </g>
        <g>
          <g>
            <path
              d="M324.267,221.867H256V153.6c0-9.426-7.641-17.067-17.067-17.067s-17.067,7.641-17.067,17.067v68.267H153.6 c-9.426,0-17.067,7.641-17.067,17.067S144.174,256,153.6,256h68.267v68.267c0,9.426,7.641,17.067,17.067,17.067 S256,333.692,256,324.267V256h68.267c9.426,0,17.067-7.641,17.067-17.067S333.692,221.867,324.267,221.867z"
            />
          </g>
        </g>
      </svg>
    </a>
  </section>
</div>
`;
